src/mpid/common/iqueue/mpidu_iqueue.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpl.h"
+#include "mpidimpl.h"
+#include "mpidu_iqueue.h"
+
+struct MPIDU_iqueue_cellblock_mgr {
+    MPL_atomic_int_t next_free_idx;
+};
+
+static void iqueue_cellblock_alloc(MPIDU_iqueue_pool_t * pool);
+static void iqueue_cellblock_free(MPIDU_iqueue_pool_t * pool, MPIDU_iqueue_cell_t * cells);
+
+int MPIDU_iqueue_pool_alloc(int cell_size, int num_cells_in_block, int max_num_cellblocks,
+                            int max_num_queues,
+                            MPIDU_iqueue_malloc_fn malloc_fn, MPIDU_iqueue_free_fn free_fn,
+                            MPIDU_iqueue_pool_t ** pool)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDU_iqueue_pool_t *new_pool = NULL;
+    uintptr_t total_queue_size = max_num_queues * sizeof(MPIDU_iqueue_queue_t);
+    uintptr_t total_cell_size = cell_size * num_cells_in_block * max_num_cellblocks;
+    void *slab = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_POOL_ALLOC);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_POOL_ALLOC);
+
+    new_pool = MPL_malloc(sizeof(MPIDU_iqueue_pool_t), MPL_MEM_OTHER);
+    if (new_pool == NULL) {
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+    new_pool->cell_size = cell_size;
+    new_pool->num_cells_in_block = num_cells_in_block;
+    new_pool->max_num_cellblocks = max_num_cellblocks;
+    new_pool->max_num_queues = max_num_queues;
+
+    new_pool->malloc_fn = malloc_fn;
+    new_pool->free_fn = free_fn;
+    slab = malloc_fn(total_queue_size + sizeof(MPIDU_iqueue_pool_t)
+                     + total_queue_size + sizeof(MPIDU_iqueue_cellblock_mgr_t)
+                     + total_cell_size);
+    if (slab == NULL) {
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+    new_pool->slab = slab;
+    new_pool->all_queues = slab;
+    new_pool->cellblk_mgr = slab + total_queue_size;
+    new_pool->all_cells = new_pool->cellblk_mgr + sizeof(MPIDU_iqueue_cellblock_mgr_t);
+    new_pool->cells = NULL;
+
+    MPL_atomic_store_int(&new_pool->cellblk_mgr->next_free_idx, 0);
+
+    *pool = new_pool;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_POOL_ALLOC);
+    return mpi_errno;
+  fn_fail:
+    if (new_pool) {
+        MPL_free(new_pool);
+    }
+    if (slab) {
+        free_fn(slab);
+    }
+    goto fn_exit;
+}
+
+void MPIDU_iqueue_pool_free(MPIDU_iqueue_pool_t * pool)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+
+    if (pool->slab) {
+        pool->free_fn(pool->slab);
+    }
+    MPL_free(pool);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+}
+
+void MPIDU_iqueue_queue_alloc(MPIDU_iqueue_pool_t * pool, int index, MPIDU_iqueue_queue_t ** queue)
```

Comment:
This function actually does not allocate any memory just returns the pointer to the queue having requested index. Should this be renamed `MPIDU_iqueue_get_queue` or `MPIDU_iqueue_init_queue`?